apps/editing-toolkit/editing-toolkit-plugin/block-patterns/class-block-patterns.php

```diff
@@ -82,75 +97,50 @@ public function register_patterns() {
 	 *
 	 * @return array
 	 */
-	public function get_patterns() {
-		$patterns_dir = __DIR__ . '/patterns/';
-		$patterns     = array();
-
-		if ( ! is_dir( $patterns_dir ) ) {
-			return $patterns;
-		}
-
-		$directory_handle = opendir( $patterns_dir );
-		if ( ! $directory_handle ) {
-			return $patterns;
-		}
-
-		// Get all the pattern files.
-		$pattern       = readdir( $directory_handle );
-		$pattern_files = array();
-		while ( false !== $pattern ) {
-			// Only allow files ending in .json or .php.
-			if ( substr( $pattern, -5 ) === '.json' || substr( $pattern, -4 ) === '.php' ) {
-				$pattern_files[] = $pattern;
+	private function get_patterns() {
+		$block_patterns = get_transient( $this->patterns_cache_key );
+
+		// Load fresh data if we don't have any patterns.
+		if ( false === $block_patterns || ( defined( 'WP_DISABLE_PATTERN_CACHE' ) && WP_DISABLE_PATTERN_CACHE ) ) {
+			$request_url = add_query_arg(
+				array(
+					'site'     => $this->source_site_id,
```

Comment:
We don't need to specify the source site for dotcompatterns, by default the endpoint will pull from that source site, the idea being that consumers of the endpoint shouldn't need to know about the source site unless there's a reason to (like for a particular starter design).

We _do_ need to specify categories for filtering, though, so to just pull the patterns without layouts, we might want to add `'categories' => 'pattern'` here.

Reply:
Can we call this something other than categories? It is probably going to get confusing with actual pattern categories like "Buttons", "Call to Action" etc. Maybe "types"?

Reply:
@apeatling good question — I'll just do some thinking out loud here:

In general, I think that anything that is technically a category on the source site (whether `dotcompatterns` or another site) should be returned by the `categories` param, which maps nicely to how the [Posts API endpoint](https://developer.wordpress.org/rest-api/reference/posts/) works.

That said, here we need a way to distinguish between the types of pattern, and the source site is currently using categories to do this. And it does make it harder if the consumer has to worry about category filtering, and excluding some categories rather than others. So, I think I'm convinced that the API endpoint should support this, and `type` does seem like a natural name for this, even though traditionally that would map to _post_ type. But it _does_ make sense to re-use it.

I'll have a play with it, and see how I go, but if you specific ideas about how it should work, let me know, a comment on pbxlJb-oL-p2 is great for visibility if there's specific behaviour you have in mind. Right now, I'd like to try out, can we treat certain categories as 'controlled' in the sense that we hide the category names from the output of the API endpoint, and just use them for filtering. For now those controlled categories would be `layout` and `pattern`, so I assume we're happy to never either of these as drop-down categories in the pattern inserter.

Reply:
@apeatling I did a bit more exploration into this, this afternoon, and I'm not sure we need this additional `type` if instead we make our categories _child_ categories of the type we want to filter by. For example, if we create all of our pattern categories in dotcompatterns and select Pattern as the _parent_ category, then when we call the endpoint with `?categories=pattern` we will get all of the patterns that are children of that category, and the parent category will not appear on each individual pattern, so you can still safely iterate over the categories there.

I've set up an example on the test site `andysfakeblockpatternsourcesite`. If you call `https://public-api.wordpress.com/rest/v1/ptk/patterns/?site=andysfakeblockpatternsourcesite.wordpress.com&categories=pattern` you get a list of two patterns, each with a different category (one has the category of `paragraph`, the other has the category of `call-to-action`), however the parent category (`pattern`) does not appear in the response.

This seems to use the default behaviour of WP_Query for filtering, and means we can avoid having to add conditional logic at the endpoint or in the Editing Toolkit plugin. Would this approach (hierarchical categories) resolve this? CC: @ianstewart as it might affect the re-categorisation (having categories be children of Pattern or Layout). And thanks @ramonjd and @lsl for the discussion, I can't remember which, but I know one of you suggested the idea of hierarchical categories as a way of avoiding locking the complexity in code.

Reply:
Just following up from 91-gh-view-design#issuecomment-701107602, we should now be able to update this to use `'tags' => 'pattern'` to grab all patterns, e.g. using the call `https://public-api.wordpress.com/rest/v1/ptk/patterns/?tags=pattern` without needing an additional query param. Does that work for you?

Reply:
Thanks for going through all of this, it all makes total sense and I think that's a great approach. :+1: I'm in the process of updating this now.